### PR TITLE
[FIXED JENKINS-28693] docker-java shading

### PR DESCRIPTION
**Pull request by oleg-nenashev**
https://issues.jenkins-ci.org/browse/JENKINS-28693

The PR has been built on the top of #16 and #17 to take new files into the account

Change to review: https://github.com/oleg-nenashev/docker-traceability-plugin/commit/2bfc689e32adb054350ec17f454bd178ba3907c7

Summary:
- [x] Modify the project structure
- [x] Create a shaded docker-java shaded with class relocation
- [x] Update the code

The proposed version works well with docker-build-step-plugin

@reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@stephenc @aheritier @jtnord Your review would be useful

**Comment by jtnord** [comment]
:-1:  you _should_ not need to an another module to shade docker-java or to rename class imports.

The glassfish and google shaded names will likely conflict with some other plugin (docker-java?) and should also be renamed to a subset of your package.

**Comment by oleg-nenashev** [comment]
@jtnord 
I've implemented another module according to the clarification of issues in maven-hpi plugin limitations in docker-plugin pull requests. I'm waiting for @aheritier to comment the approach

**Comment by jtnord** [comment]
if you are referring to [this commit](https://github.com/jenkinsci/docker-plugin/commit/cbbc8a05d228981303ede1f460bb428fd75a42ed) then shade plugin was not configured with an execution.

**Comment by jglick** [comment]
At root a bug in the client library. Suggested a simpler workaround in the issue.

**Comment by oleg-nenashev** [comment]
@jglick 
Thanks for the proposal. Patching of docker-java could help in the case of docker-build-step, but for us the shading is a problem in any case if somebody has an old version, because installation of the new plugin may break the installation. I also doubt we can timely get this fix released and then test everything else.

What do we do with this change?

**Comment by oleg-nenashev** [comment]
> if you are referring to this commit then shade plugin was not configured with an execution.

Follow-up to the discussion in HipChat. There's a rumor shading does not work together with maven-hpi

**Comment by jglick** [comment]
> Patching of docker-java could help

Well that is the long-term fix of the root issue. The short-term workaround is to override the context class loader in the Jenkins plugin.

**Comment by oleg-nenashev** [comment]
@jglick 
> The short-term workaround is to override the context class loader in the Jenkins plugin.

No idea how to do it quickly. If we decline the proposed "complex" fix, I would be grateful if somebody helps to kick-off a short-term workaround (to 3 plugins?)

**Comment by aheritier** [comment]
Here are more details about the HPI and shade Maven plugins incompatibility : http://echelog.com/logs/browse/jenkins/1427839200

I already discussed with @KostyaSha who reported the issue and asked me some help to fix it but for now I did found the time to do it. Here is the feedback he gave to me :

> The issue is that it ignores any package phase bindings and packages jars with hardcoded jar archiver. Jenkins has many issues because of classloader and it will be very convenient to have ability to execute shad-plugin directly from hpi packaging. If packaging is jar, then maven-jar-plugin executes fine maven-shade-plugin. So my idea was to remove hardcoded jar archiver from hpi plugin and give it to maven-jar. https://github.com/jenkinsci/maven-hpi-plugin/blob/ded54b3ac24035ebacda64a0900c3dea5e3de852/src/main/java/org/jenkinsci/maven/plugins/hpi/HpiMojo.java#L122-L131

File: docker-traceability-plugin/src/main/resources/org/jenkinsci/plugins/docker/traceability/fingerprint/DockerDeploymentFacet/main.jelly [hunk 1]
```diff
@@ -45,7 +45,7 @@
   </h3>
   <ul>  
     <li>${%Image ID}: <docker:image id="${it.imageId}"/></li>
-    <li>${%Image name}: &quot;${report.imageName}&quot;
+    <li>${%Image name}: &quot;${report.imageName}&quot;</li>
```
Comment (oleg-nenashev) [inline docker-traceability-plugin/src/main/resources/org/jenkinsci/plugins/docker/traceability/fingerprint/DockerDeploymentFacet/main.jelly]: This minor thing sneaked from another change (API cleanup). NIT IMO

**Comment by oleg-nenashev** [comment]
@jtnord @jglick @aheritier 
Ready for the final review. I've fixed previous comments and merged with the master branch

**Comment by jglick** [comment]
:+1: though as time permits please submit an upstream patch correcting the misuse of `ServiceLoader`, and you can check whether a temporary override of `Thread.currentThread().contextClassLoader` works around the issue.

File: docker-java-shaded/pom.xml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  <modelVersion>4.0.0</modelVersion>
+
+  <!-- nb: No parent, otherwise we inherit wrong versions from
+  transient jenkins dependencies. -->
+
+  <groupId>org.jenkins-ci.plugins</groupId>
+  <artifactId>docker-java-shaded-traceability</artifactId>
+  <packaging>jar</packaging>
+  <version>1.0-beta-4-SNAPSHOT</version>
+  <name>docker-java shaded for Cloudbees Docker Traceability</name>
```
Comment (KostyaSha) [inline docker-java-shaded/pom.xml]: Bees from upper letter

File: docker-traceability-plugin/src/main/java/org/jenkinsci/plugins/docker/traceability/DockerTraceabilityPlugin.java [hunk 3]
```diff
@@ -111,3 +111,4 @@ public void start() throws Exception {
         load();
     }
 }
+
```
Comment (KostyaSha) [inline docker-traceability-plugin/src/main/java/org/jenkinsci/plugins/docker/traceability/DockerTraceabilityPlugin.java]: redundant new line?

**Comment by KostyaSha** [comment]
> if you are referring to this commit then shade plugin was not configured with an execution.

o'rly? Could you show your working variant of shading with hpi plugin?

File: docker-java-shaded/pom.xml (hunk 2 above)
Comment (oleg-nenashev) [inline docker-java-shaded/pom.xml]: Thanks!

**Comment by oleg-nenashev** [comment]
> o'rly? Could you show your working variant of shading with hpi plugin?

See the comment from @aheritier above. We agreed that the shading need's some adjustments. I've created https://issues.jenkins-ci.org/browse/JENKINS-28796

**Comment by jtnord** [comment]
> ..ker/traceability/model/DockerRecordsRegistry.java → ...ker/traceability/model/DockerRecordsRegistry.java

Thanks github - I shall remember to buy my chocolate fireguard from you as well!

I am assuming (!) that there is nothing in these renames other than moving from root to the new (correct) submodule.

based on that assumption :+1:

**Comment by oleg-nenashev** [comment]
> I am assuming (!) that there is nothing in these renames other than moving from root to the new (correct) submodule.

Yes, folder moves only

**Comment by KostyaSha** [comment]
imho this is waste of time/resources, history trashing, etc.  Better enhance maven-hpi plugin.

File: docker-java-shaded/pom.xml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            <relocation>
+              <pattern>org.apache.http</pattern>
+              <shadedPattern>shaded.org.apache.http</shadedPattern>
+            </relocation>
+            <relocation>
+              <pattern>com.google.common</pattern>
+              <shadedPattern>shaded.com.google.common</shadedPattern>
+            </relocation>
+            <relocation>
+              <pattern>com.github.dockerjava</pattern>
```
Comment (KostyaSha) [inline docker-java-shaded/pom.xml]: I had no time for cleaning and this was copy-pasted to this plugin `<pattern>com.github.dockerjava</pattern>` vs `<include>com.github.docker-java*:*</include>`

**Comment by KostyaSha** [comment]
Ah, already merged.

**Comment by oleg-nenashev** [comment]
That's why we have a beta-5 opportunity :)

**Comment by oleg-nenashev** [comment]
@KostyaSha 
Everything is correct BTW. "include" takes artifact IDs and "relocation/pattern" takes package prefixes

**Pull request merged**